### ISSUE TITLE
Installer

### ISSUE THREAD

**Issue by janmechtel** [janmechtel/rockettype#2]
Would be really nice if one could download an .exe or .msi to install everything.

## Installer candidates

Which installer should we use?
Apparently we need something to included the Python stuff and then something to create an installer.

### Combined Freeze and Installer:
* **[pynsist](https://pynsist.readthedocs.io/en/latest/faq.html#alternatives) - Favorite**, see test below
* [Cyton](https://stackoverflow.com/questions/2581784/can-cython-compile-to-an-exe)

### Freeze Tools:

* [PyInstaller](https://www.pyinstaller.org/) - see test below
* [WinPython + InnoSetup](https://cyrille.rossant.net/create-a-standalone-windows-installer-for-your-python-application/)
* [cx_Freeze](http://cx-freeze.sourceforge.net/) 
* Python compilers like [Nuitka](http://nuitka.net/).

### Actual Installers:
* InnoSetup
* (Create a single .exe) > extracts on each load
* MSIX?
* Windows Installer

## What to consider:
* [Size](https://itnext.io/python-guis-with-dearpygui-137f4a3360f2)
* Usability
* Does it create a .exe (or only a shortcut to launch?)
* Can in build for windows from other OS
* [Pyninst considerations](https://pynsist.readthedocs.io/en/latest/faq.html#alternatives)
* [Some comments on the topic](https://news.ycombinator.com/item?id=24319135)

## Signing

This will probably trigger [Windows Smartscreen / Defender](https://docs.microsoft.com/en-us/windows/security/threat-protection/microsoft-defender-smartscreen/microsoft-defender-smartscreen-overview) warnings when publishing unsigned / without and EV certificate. Even with the certificate there will be a warning for the first couple of downloads.

See https://pynsist.readthedocs.io/en/latest/faq.html#code-signing

One way around this would be to go through the [Windows Store](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-distribute)

* Can key loggers be distributed through the app store? >> Let's find out :)
* Can python programs be distributed through the app store? >> Probably, but first we need an .exe and the convert it to .msix ([see here](https://techcommunity.microsoft.com/t5/msix-packaging-and-tools/unable-to-create-a-python-script-to-msix/m-p/1180343))

Windows Store would also provide some basic stats?

**Comment by janmechtel** [comment]
# PyInstaller

`pip install pyinstaller`
`pyinstaller logger.pyw`

- Worked. 
- Created folder called "logger" with logger.exe
- Size: 29.5 MB
- Python required: No (untested, [Source](https://pyinstaller.readthedocs.io/en/stable/))
- Build Windows on other OS: No
- Installer included: No (? still needs InnoSetup or similar?)

## TODO
- Installer?
- Test (Does this really include all the python needed?)
- How to change the name?
- Can we include the stats calculation in the same thing?

**Comment by janmechtel** [comment]
# pynist

## Installation
1. https://nsis.sourceforge.io/Download (3.06.1)
2. pip install pynsist
3. Write `installer.cfg` 
4. `python -m nsist installer.cfg`

## Result
- Worked. 
- Created folder called built with RocketType_0.1.exe which is a complete installer
- Size: 16 MB
- Python required: No 
- Build Windows on other OS: Yes
- Installer included: Yes NIST

## TODO
- Installation requires a lot of buttons, can that be avoided by NSIST settings?
- As we are running the script the notification still says "Python" #3 
- Exiting worked, but didn't show the notification reliably?
- What about the stats?

**Comment by janmechtel** [comment]
Not working yet :-(
The RocketType Start menu shortcut contains this:

`C:\Users\JanMechtel\AppData\Local\Programs\RocketType\Python\pythonw.exe "C:\Users\JanMechtel\AppData\Local\Programs\RocketType\logger.pyw"`

Which doesn't seem to work. I tricked myself into thinking it works by double clicking logger.pyw

**Comment by janmechtel** [comment]
running with python.exe instead `C:\Projects\Keys\RocketType\build\nsis\Python\python.exe "C:\Projects\Keys\RocketType\build\nsis\logger.pyw"` 

1. Revealed that "import six" fails, adding `six==...` to the .cfg fixed that.
2. Now the next one is pkg_resources which seems to be related to setuptools addin to .cfg fixed that
3. win32api missing added pypiwin32 didn't work

... the last one was tricky

"C:\Users\JanMechtel\AppData\Local\Programs\Python\Python38\lib\site-packages\nsist\wheels.py", line c. 40,

remove the `m` from `d = {'cp%sm' % py_version_nodot: 3,  # Is the m reliable?` and turn it into   `d = {'cp%s' % py_version_nodot: 3,  # Is the m reliable?` 

then it at least builds the installer but still fails :(

I've confirmed that there is a file `pkgs/win32/win32api.pyd` .pyd is supposed to be a windows specific python module.

I can't import it.

I tried runnning only the nsist/python.exe and import there.

import pynput
print(pynput)

works

also import sys and print(sys.path) shows the correct path

I'm a bit out of luck why the module can not be loaded I'll ask on StackOverflow maybe it's related to 32 vs. 64 bit (however import platform + platform.architecture shows 64bit both in the nsist and my normal local version.

Otherwise we could try:
- another notification module than win10toast or
- another freezer+package

**Comment by janmechtel** [comment]
Partial success the nsist\python.exe can `import win32api` after:

```
sys.path.append('C:\\Projects\\Keys\\RocketType\\build\\nsis\\pkgs\\win32\\lib')
sys.path.append('C:\\Projects\\Keys\\RocketType\\build\\nsis\\pkgs\\win32')
import pywin32_bootstrap
import win32api
```

Now that I found the problem I fixed it and submitted a PR https://github.com/takluyver/pynsist/pull/204

**Comment by janmechtel** [comment]
@appins can you please try the installer from the homepage and report back?

**Comment by appins** [comment]
@janmechtel I tried this, the program worked the first time and then stopped working. Running it from the command line yields no output. I suspect it might have trouble trying to capture the keyboard output the second time around. Could you see if you need to close access to any of the files/objects you're using in the code?

**Comment by appins** [comment]
Also, you should consider cython, it might be nice to have a portable binary, as this would also solve #12

**Comment by janmechtel** [comment]
> 
> 
> @janmechtel I tried this, the program worked the first time and then stopped working. Running it from the command line yields no output. I suspect it might have trouble trying to capture the keyboard output the second time around. 

Can you:

- [x] Check task manager if python.exe or pythonw.exe are still running?
- [x] inspect the startmenu shorctut and run it with python.exe instead of pythonw.exe to make sure there are no exceptions.

I will:

- [x] if we need to close access to any of the files/objects we're using in the code

**Comment by janmechtel** [comment]
I don't have any idea about objects we are blocking, even the listener is not exclusive. In fact i can start the installed app multiple times at once ...

I will:

- [x] Ask a friend to see if the problem exists for him

**Comment by janmechtel** [comment]
# Cython

I'm not really sure how this would work yet and believe this would also create one giant .exe not a real installer

https://stackoverflow.com/questions/2581784/can-cython-compile-to-an-exe

**Comment by appins** [comment]
Yeah, I tried opening task manager and python and pythonw were both running. When I killed the tasks and reopened it, it continued to not work and the processes reappeared. I will try restarting my computer since maybe that has something to do with it and get back to you. 

As for Cython, it would create a large executable, but maybe that wouldn't be too bad of a way to use the program. I feel like people might be more willing to have a keylogger be a portable program rather than an installer, but maybe that's a project for a later time.
